File: src/main/java/org/featx/spec/model/ListResponse.java
package org.featx.spec.model;

import org.featx.spec.constant.ErrorCode;
import org.featx.spec.enums.BusinessError;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

import java.util.List;
import java.util.Optional;

/**
 * @author Excepts
 * @since 2019/10/27 11:01
 */
@NoArgsConstructor
@EqualsAndHashCode(callSuper = true)
public class ListResponse<R> extends BaseResponse<List<R>> {

    private static final long serialVersionUID = -1370862431440698540L;

    private ListResponse(int code, String message, List<R> result) {
        this.setCode(code);
        this.setMessage(message);
        this.setData(result);
    }

    public Integer getSize() {
        return Optional.ofNullable(getData()).map(List::size).orElse(0);
    }

    public static <R> ListResponse<R> succeeded(List<R> result) {
        return new ListResponse<>(ErrorCode.NO_ERROR, null, result);
    }

    public static <R> ListResponse<R> occur(BusinessError businessError, List<R> extra) {
        return new ListResponse<R>(businessError.getCode(), businessError.getMessage(), extra);
    }
}

File: src/main/kotlin/org/featx/jvmon/model/ListResponse.kt
package org.featx.jvmon.model

import org.featx.jvmon.exception.BusinessError
import org.featx.jvmon.exception.ErrorCode
import java.util.*

/**
 * @author Excepts
 * @since 2019/10/12 11:01
 */
open class ListResponse<T>(code: Int, msg: String?, result: List<T>?) :
        BaseResponse<List<T>>(code, msg, result) {

    private val size: Int?
        get() = Optional.ofNullable(getResult())
                .map<Int> { t -> t.size }
                .orElse(0)

    companion object {
        fun <R> succeeded(result: List<R>): ListResponse<R> {
            return ListResponse(ErrorCode.NO_ERROR, null, result)
        }

        fun <R> occur(businessError: BusinessError): ListResponse<R> {
            return ListResponse(businessError.code, businessError.message, null)
        }

        fun <R> occur(businessError: BusinessError, extra: List<R>): ListResponse<R> {
            return ListResponse(businessError.code, businessError.message, extra)
        }
    }
}